doc/adr/0003-distributed-tracing.md

```diff
@@ -0,0 +1,158 @@
+existing operations tooling in use by our users. Additionally Grafana 7 supports Jaeger and Zipkin today.  
+
+Long term I think every server that handles a message should emit a unique trace so we can also get visibility into
+the internal flow of the NATS system and exactly which gateway connection has a delay - see our current HM issues - but
+ultimately I don't think we'll be doing that in the hot path of the server, though these traces are easy to handle async
+
+Meanwhile, this proposal will let us get very far with our current Latency Advisories.
+
+## Configuring an export
+
+Today there are no standards for the HTTP headers that communicate span context downstream - with Trace Context being
+an emerging w3c standard.
+
+I suggest we support the Jaeger and Zipkin systems as well as Trace Context for long term standardisation efforts.
+
+Supporting these would mean we have to interpret the headers that are received in the request to determine if we should
+publish a latency advisory rather than the static `50%` configuration we have today.
+
+Today we have:
+
+```
+exports: [
+    {
+        service: weather.service 
+        accounts: [WEB]
+        latency: {
+            sampling: 50% 
+            subject: weather.latency
+        }
+    }
+]
+```
+
+This enables sampling based `50%` of the service requests on this service.
+
+I propose we support additional sampling values `zipkin`, `jaeger`, `trace_context` which will configure the server to
+interpret the headers as below to determine if a request should be sampled.
+
+## Propagating headers
+
+The `io.nats.server.metric.v1.service_latency` advisory gets updated with additional `trace_format` and `headers` fields.
+
+`headers` would just propagate all the headers unmodified. We might later add support for a whitelist here to avoid
+leaking sensitive information.
+
+```json
+{
+  "type": "io.nats.server.metric.v1.service_latency",
+  "id": "YBxAhpUFfs1rPGo323WcmQ",
+  "timestamp": "2020-05-21T08:06:29.4981587Z",
+  "status": 200,
+  "trace_format": "jaeger",
+  "headers": {
+    "Uber-Trace-Id": "09931e3444de7c99:50ed16db42b98999:0:1"
+  },
+  "requestor": {
+    "acc": "WEB",
+    "rtt": 1107500,
+    "start": "2020-05-21T08:06:20.2391509Z",
+    "user": "backend",
+    "lang": "go",
+    "ver": "1.10.0",
+    "ip": "172.22.0.7",
+    "cid": 6,
+    "server": "nats2"
+  },
+  "responder": {
+    "acc": "WEATHER",
+    "rtt": 1389100,
+    "start": "2020-05-21T08:06:20.218714Z",
+    "user": "weather",
+    "lang": "go",
+    "ver": "1.10.0",
+    "ip": "172.22.0.6",
+    "cid": 6,
+    "server": "nats1"
+  },
+  "start": "2020-05-21T08:06:29.4917253Z",
+  "service": 3363500,
+  "system": 551200,
+  "total": 6411300
+}
+```
+ 
+## Header Formats
+
+Numerous header formats are found in the wild, main ones are Zipkin and Jaeger and w3c `tracestate` being an emerging standard.
```

Comment:
If they are using these products then they do not need to use our latency metrics for tracking as well no?

I think where this can be useful is in mostly transparent scenarios where they are not using client side libraries but want latency measurements to also drive other systems.

Reply:
The effort here is to enable us to place our metrics next to their existing infrastructure metrics.

Surveyor would receive these advisories and publish them to jaeger appropriately, this is about interop with other tools/systems

Reply:
Understood. Question is do they use the tracing libs in the app as well, that was what I was trying to get at.

Reply:
they could have the nats service/responder also emit spans if they want, the app would have access to these headers and can inject things into the trace in addition.

For the other parts I showed in my image attachd to the PR traefik was just a flag to enable it and in the go web micro service I added some code:

```
package main

import (
	"bytes"
	"fmt"
	"log"
	"net/http"
	"os"
	"time"

	"github.com/nats-io/nats.go"
	"github.com/opentracing/opentracing-go"
	"github.com/opentracing/opentracing-go/ext"
	"github.com/uber/jaeger-client-go"
	jaegercfg "github.com/uber/jaeger-client-go/config"
	"github.com/uber/jaeger-lib/metrics"
)

var nc *nats.Conn

func connect() (err error) {
	if os.Getenv("NATS_URL") == "" {
		return fmt.Errorf("NATS_URL is required")
	}

	opts := []nats.Option{
		nats.ErrorHandler(func(_ *nats.Conn, _ *nats.Subscription, err error) {
			panic(err)
		}),
	}

	if os.Getenv("NATS_USER") != "" && os.Getenv("NATS_PASSWORD") != "" {
		opts = append(opts, nats.UserInfo(os.Getenv("NATS_USER"), os.Getenv("NATS_PASSWORD")))
	}

	nc, err = nats.Connect(os.Getenv("NATS_URL"), opts...)

	return err
}

func weatherHandler(w http.ResponseWriter, r *http.Request) {
	tracer := opentracing.GlobalTracer()

	// extract the incoming trace details from the header, if headers
	// are there and specifically set this trace to unpublished follow
	// ones would also not be instrumented
	spanCtx, _ := tracer.Extract(opentracing.HTTPHeaders, opentracing.HTTPHeadersCarrier(r.Header))
	serverSpan := tracer.StartSpan(r.Method+` `+r.RequestURI, ext.RPCServerOption(spanCtx), ext.SpanKindProducer)
	defer serverSpan.Finish()

	log.Printf("Handling weather request: %v", r)

	resp, err := nc.Request("weather.service", []byte{}, time.Second)
	if err != nil {
		w.WriteHeader(http.StatusBadGateway)
		w.Write([]byte(err.Error()))
		serverSpan.SetTag("error", true)
		serverSpan.SetTag("message", err.Error())
		return
	}

	w.Write(resp.Data)
}

func main() {
	err := connect()
	if err != nil {
		panic(err)
	}

	cfg := jaegercfg.Configuration{
		ServiceName: "http-backend",
		Sampler: &jaegercfg.SamplerConfig{
			Type:  jaeger.SamplerTypeConst,
			Param: 1,
		},
		Reporter: &jaegercfg.ReporterConfig{
			LocalAgentHostPort: fmt.Sprintf("%s:%s", os.Getenv("JAEGER_AGENT_HOST"), os.Getenv("JAEGER_AGENT_PORT")),
		},
	}

	tracer, closer, err := cfg.NewTracer(jaegercfg.Metrics(metrics.NullFactory))
	if err != nil {
		panic(err)
	}
	defer closer.Close()
	opentracing.SetGlobalTracer(tracer)

	log.Printf("Listening on :8080 for /weather")
	http.HandleFunc("/weather", weatherHandler)
	log.Fatal(http.ListenAndServe(":8080", nil))
}

```